tools/codegen/src/ir/mod.rs

```diff
@@ -53,6 +53,7 @@ pub(crate) struct Option_ {
 pub(crate) struct Union {
     name: String,
     items: Vec<ItemDecl>,
+    custom_item_ids: Option<Vec<usize>>,
```

Comment:
can we change the items definition instead of adding an option field?
```
items: Vec<UnionItemDecl>

UnionItemDecl {
  typ: String, 
  id: usize
}
```

I think this change will simplify the ir and code generation related code

Reply:
I added an option field `custom_item_ids` instead of change the items definition beacuse I want back compatibility:


Beacuse the `v0.7.3` `ItemDecl` is `serde(transparent)`, for example, a union instance will be serialized into this:
```rust
    let union = Union {
        name: "Union".to_string(),
        items: vec![ItemDecl {
            typ: "u8".to_string(),
        }],
        imported_depth: 0,
   };
   // union's serialization result will be:
{
      "type": "union",
      "name": "SyncMessage",
      "items": [
        "a0",
        "a1",
        "a2"
      ]
}
```

If I change the items's definition to
```rust
items: Vec<UnionItemDecl>

UnionItemDecl {
  typ: String, 
  id: usize
}
#[derive(Debug, Property, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct NewUnion {
    name: String,
    items: Vec<UnionItemDecl>,
    #[serde(default = "zero", skip_serializing_if = "is_zero")]
    imported_depth: usize,
}
```

then the `NewUnion`'s serialization result will be
```rust
{
      "type": "union",
      "name": "SyncMessage",
      "items": [
	     {"typ":"a0",id:0},
	     {"typ":"a1",id:1},
	     {"typ":"a2",id:2},
      ]
}

```

then the `v0.7.3` 's ir json result won't be deserialized into  `NewUnion` struct

Reply:
can it be resolved by a customized deserializer?
```rust
pub(crate) struct NewUnion {
    name: String,
    #[serde(deserialize_with = "from_combatible_vec")]
    items: Vec<UnionItemDecl>,
    #[serde(default = "zero", skip_serializing_if = "is_zero")]
    imported_depth: usize,
}
```

Reply:
I think I can do it, and I'm going to try this by a customized deserializer.

Reply:
https://github.com/nervosnetwork/molecule/pull/65/commits/973021007d90a36d0b6afff390c2f4ba3f50ac5a

The latest commit implement customize deserializer. :laughing: